### DISCUSSION TITLE
Ist beabsichtigt, dass sich die beiden Suchfelder unterschiedlich verhalten?

### DISCUSSION THREAD

**Discussion by matthias-ronge** [kitodo/kitodo-production#5836] [Q&A]
Nachdem ich beim Praxistreffen die Frage gestellt hatte, wie es möglich sei, nach METS-Dateiinhalten zu suchen. Man kann diese nicht finden, wenn man im Vorgangs-Filterschlitz sucht. @oliver-stoehr hat weiter festgestellt, dass man jedoch danach suchen kann, wenn man im Suchfeld oben auf der Seite (im blauen Bereich) sucht.

**Suche nach einem String aus den Metadaten im Filter:** (kein Treffer)

![01_filter_metadata](https://github.com/kitodo/kitodo-production/assets/3040657/131cdf18-8cbd-43f9-9906-56b1ac6cd541)

**Suche nach einem String aus den Metadaten im oberen Suchschlitz:** (1 Treffer)

![02_suche_metadata](https://github.com/kitodo/kitodo-production/assets/3040657/a732b411-fe46-4ee3-a9fb-00432a643546)

Ich habe mir das etwas weiter angesehen und festgestellt, dass sich der Suchschlitz auch sonst etwas anders verhält. Wenn man nach einer ID sucht, funktioniert es dort, während es im Filter-Schlitz nur funktioniert, wenn man explizit das Präfix „id:“ davor stellt. Wenn man das Präfix „id:“ in den oberen Suchschlitz eingibt, bekommt man hingegen einen Stacktrace.

**Suche nach „7517“ im Filter:** (kein Treffer)

![03_filter_vnr_ohne_id](https://github.com/kitodo/kitodo-production/assets/3040657/31c6d4a5-e63c-4ae6-bdba-312aec9da68b)

**Suche nach „id:7517“ im Filter:** (1 Treffer)

![04_filter_vnr_mit_id](https://github.com/kitodo/kitodo-production/assets/3040657/52f6d1e4-a58f-4214-bebb-af675760fd3b)

**Suche nach „7517“ im oberen Suchschlitz:** (1 Treffer)

![05_suchschlitz_vnr_ohne_id](https://github.com/kitodo/kitodo-production/assets/3040657/71890a10-3957-46b6-b2a1-b681f5b7e2ba)

**Suche nach „id:7517“ im oberen Suchschlitz:** (stürzt ab)

![06_suchschlitz_vnr_mit_id](https://github.com/kitodo/kitodo-production/assets/3040657/cdcd6bc4-e9e2-43ae-9858-7edbcb80200c)

Auf den zweiten Blick fallen auch Unterschiede in der Seite auf, je nachdem, über welches Suchfeld man sucht. Am Auffälligsten sind:
- die etwas aus dem Layout gefallenen weiteren Filtermöglichkeiten am unteren Ende der Seite (nur über oberen Suchschlitz verfügbar)
- dass abgeschlossene Vorgänge einmal  standardmäßig gefunden werden, einmal nicht
- der Filter-Suchschlitz penetrant ein Speichern des Filters aufdrängt, der obere Suchschlitz bietet die Möglichkeit hingegen gar nicht an

**Filter:**

![unterschiede_filter](https://github.com/kitodo/kitodo-production/assets/3040657/9802f228-c059-498d-bbe8-deb63a28d4f4)

**Oberer Suchschlitz:**

![unterschiede_suche](https://github.com/kitodo/kitodo-production/assets/3040657/ef59727f-27a8-4a4e-9c38-ffd2582a8915)

**Meine Frage, auch an das @kitodo/kitodo-community-board, ist: Ist dieses Verhalten 2023 so noch gewollt?** Meine persönliche Erwartungshaltung wäre:

- Beide Suchschlitze verhalten sich gleich. (Wird der obere Suchschlitz auf einer anderen Seite als der Vorgangsliste benutzt, sollte die Ansicht zur Vorgangsliste wechseln.)
- Die Suche nach einer ID ist ohne den Prefix „id:“ möglich
- Die Suche nach Metadatenwerten ist möglich
- Die Suche mit dem Prefix „id:“ stürzt nicht ab
- Nach Benutzung des oberen Suchschlitzes wird die Suche im Filter-Suchschlitz angezeigt und kann dort gespeichert werden
- Das Speichern könnte dezenter angeboten werden (z.B. Speichern-Symbol neben dem Suchschlitz, aber nicht das Pop-Up.)
- Die zusätzlichen Suchmöglichkeiten sind grundsätzlich verfügbar
- Gleiches Verhalten hinsichtlich der Anzeige abgeschlossener Vorgänge
- Trefferzahl besser positionieren

**Zum Beispiel so:**
![zusammen](https://github.com/kitodo/kitodo-production/assets/3040657/d4bbed02-542b-4df9-87e9-6ed8cf71f9b8)

**Comment by oliver-stoehr** [comment]
Vielen Dank @matthias-ronge für die ausführliche Recherche und Beschreibung!

Ich habe dazu noch folgende Ergänzungen:
- Die von dir getestete Version ist nicht mehr ganz aktuell. Es sind bereits einige UI-Verbesserungen bezüglich der Filter in Version 3.6.0 enthalten, die unter Anderem das beschriebene zu aufdringliche, automatische Öffnen des Popups beheben und die Filter-Buttons von unterhalb der Tabelle in ein neues, inhaltlich passenderes Popup an den Filtern rechts oben verschieben. (Siehe #5523)
- Eine Suche im Filterfeld ohne Präfix ist eine Suche nach dem Vorgangstitel (geht auch mit dem Präfix „process“ oder „prozess“). Daher werden dabei keine IDs gefunden. Ob das so noch sinnvoll ist oder auch IDs finden sollte, sollte hier auf jeden Fall diskutiert werden. 
- Die Unterschiede in der Ergebnisliste zwischen den beiden Such-/Filterfeldern kommt daher, dass die Ergebnisse der Filter auf beiden normalen „Vorgänge“- und „Aufgaben“-Seiten angezeigt wird. Das Suchfeld oben leitet aber auf eine separate, dritte Seite. Diese ist der „Vorgänge“-Seite sehr ähnlich. Da es aber eine separate Seite ist, wurden in der Vergangenheit nicht alle Änderungen aus der Vorgangsliste auch auf diese Seite übernommen. Aus meiner Sicht wäre es sinnvoll, wenn wir – auf welchem Weg auch immer – auf die separate Seite verzichten könnten und die Funktionen des oberen Suchfeldes auf den beiden hauptsächlich genutzten anbieten könnten.

**Comment by andre-hohmann** [comment]
@matthias-ronge und @oliver-stoehr : 

Vielen Dank für Eure Beiträge. Ich würde eine umfassende Beurteilung sehr begrüßen und will auf folgendes Issue hinweisen, das bestehende Fragen, Wünsche, ... bezüglich der Suche zusammenfasst:
- #5020 

Es muss auch darauf hingewiesen werden, dass Vieles funktioniert und einige Einrichtungen keine weiteren Anforderungen an die Suche stellen. 

Ich will jedoch niemanden davon abhalten, sich mit diesem Thema zu beschäftigen. Ich vermute jedoch, dass allein die Untersuchung und Abstimmung über die Anpassung der Suche/Filter unter den interessierten Einrichtungen sehr viel Zeit benötigt.

**Comment by matthias-ronge** [comment]
Die zwei Seiten stören mich auch, aber was mich mehr stört ist, dass dieselbe Suchsyntax in den beiden Suchschlitzen zu _unterschiedlichen_ Ergebnissen führt. (Verbunden mit der Frage, ob das inder Umstellung auf Hibernate Search wieder so umgesetzt werden soll, was umständlicher wäre, oder _eine_ Lösung für _beide_ Suchschlitze ausreichen würde. Das wäre einfacher zu implementieren, und − wie ich finde − dem Benutzer auch besser vermittelbar.)

Meine Erwartung wäre, dass eine Suche ohne ein Präfix eine „freie Suche“ ist, und z.B. mit „7517“ sowohl den Vorgang mit dieser ID findet, als auch einen Vorgang mit dem Titel _LoremIpsum_7517123X_, alle Vorgänge deren Elternteil der Vorgang mit der ID 7517 ist, und genauso auch einen Vorgang mit einem Film, bei dem im Metadatum „Länge [m]“ die Zahl 7517 steht …

**Comment by andre-hohmann** [comment]
@matthias-ronge : Entschuldige, ich ließ mich von den ausführlichen Beschreibungen hinreißen. 

Die Frage nach einer einheitlichen Syntax lässt sich natürlich deutlich einfacher abstimmen und beantworten. 
Spontan würde ich persönlich zustimmen, dass nur eine Syntax angewendet wird, wenn dies realisierbar ist. 
Es wäre natürlich besser, wenn mehr Einschätzungen gegebenen werden. Ich werde das auch in das Community Board einbringen.

**Comment by andre-hohmann** [comment]
@matthias-ronge : 
Wir haben die Frage im Community Board besprochen und stimmen grundsätzlich zu, dass nur eine Suchsyntax angewendet wird. Dies ist den Anwendern und Anwenderinnen deutlich besser zu vermitteln. 
Allerdings sind wir unsicher, zu welchen Konsequenzen diese Entscheidung führt oder führen kann. Im Folgenden wird die Diskussion skizziert, um unsere Fragen und Unsicherheiten zu vermitteln. 
Es soll vermieden werden, dass Entscheidungen auf der Grundlage falscher Vorstellungen getroffen werden. 

Die Suche darf zum Beispiel nicht auf die von Dir beschriebene Funktion reduziert werden, dass immer alles gefunden wird. Die Eingrenzung mit Präfixen, wie zum Beispiel `id`, `project`, `property`…  muss weiterhin möglich sein. Gerne auch in der Suche. 

Die unterschiedliche Suchsyntax entstand unseres Wissens dadurch, weil in den Filtern und der Erweiterten Suche die Suchsyntax der Eigenschaften aus Kitodo.Production 2 übernommen wurde und im Suchfeld der ElasticSearch-Index abgefragt wird, der die Präfixe nicht "kennt". 

Die Diskussion im Community Board wurde häufig auf weitere Fragen und Themen gelenkt, die in diesem Zusammenhang berücksichtigt werden sollten, wie zum Beispiel: 
- Perspektive der Vorgangseigenschaften (die eigentlich entfernt werden sollten)
- Informationen darüber, was im Index enthalten ist (Metadaten METS-Datei, …)
- Informationen darüber, welche Auswirkungen Hibernate Search hat
- Erwartungen an die Suche, insbesondere spezifischere Suchen über die Metadaten der METS-Datei

Um Fortschritte zu erzielen, sollten jedoch folgende Aspekte unterscheiden werden: 
1. Kurzfristig: Rasche Beantwortung Deiner Fragen, so dass Du weiterarbeiten kannst. 
2. Mittelfristig: Abstimmung über das Konzept und die Ziele des Projekts, um mögliche weitere Fragen beantworten zu können. 

Bezüglich 1 würden wir mit unserem derzeitigen Kenntnisstand eine einheitliche Suchsyntax befürworten. Insbesondere, wenn dies die Umsetzung vereinfacht. 
Bezüglich 2 würden wir gerne ein Treffen vereinbaren, um das Konzept (besser) verstehen zu können - wenn möglich zusammen mit @solth und @oliver-stoehr. Du hast ja selber auf weitere Verbesserungsmöglichkeiten hingewiesen, wie zum Beispiel die Ergebnisliste der Suche. 

@kitodo/kitodo-community-board : 
Ich hoffe, dass der Verlauf/die Ergebnisse der Diskussion im Community Board korrekt beschrieben sind. Wenn nicht, bitte gerne ergänzen oder korrigieren.

**Comment by subhhwendt** [comment]
Nach einem Austausch vom @kitodo/kitodo-community-board mit @matthias-ronge,  @solth  und @oliver-stoehr  halte ich von dieser Videokonferenz fest - bitte bei Bedarf ergänzen / kommentieren.

- Es wird eine Integration der sog. Headersuche (Suchschlitz oben links mit einer Metadatensuche im Index auf Basis eines aktuellen Standes der meta.xml) in den Filtersuchschlitz angestrebt (metadata:....). 
- Aktuell lässt sich in der sog. Headersuche nur übergreifend nach Elementen aus den Metadaten suchen ohne Einschränkung auf ein bestimmtes Metadatenfeld. In einer Ausbaustufe sollte eine Umsetzung von gezielteren Suchen mit key:value-Abfragen angestrebt werden. Um bei dieser Ausbaustufe nicht mit den evtl. wenig sprechenden internen Namen aus meta.xml zu operieren wurde u.a. eine Verankerung von sprechenderen Namen im Ruleset kurz andisktutiert.
- Mit dieser Integration wird die Headersuche oben links überflüssig. Die für die Headersuche entwickelte sehr spezielle Trefferliste mit anzuwählenden Filtermöglichkeiten unter der Trefferliste wird ebenfalls aufgegeben.
- Die einmalige Aufnahme von Metadaten beim Import in die sog. properties in der DB wird ebenfalls perspektivisch aufgegeben und die damit verbundenen Abfragemöglichkeiten "property: ...".

_Angerissene Fragen für evtl. Fortsetzungsrunden:_
- Möglichkeiten der Suchsyntax abstecken. Nach der Sitzung hat @oliver-stoehr nachgetragen, dass die Filtersuche bereits "Nicht-" und "Oder-Verknüpfungen" erlaubt - siehe auch [Dokumentation im Wiki](https://github.com/kitodo/kitodo-production/wiki/Suche-und-Filter#filteralternativen-oder-verknüpfung-von-filterkriterien)
- Ist die Indexierung von vielen Datenbankinhalten sinnvoll? Wofür können sie genutzt werden? Bringt Hibernate Search die erhoffte Performance und Alltagstauglichkeit im Vergleich zur aktuellen Situation?
- Die aktuelle Indexierung der Metadaten ist mit seiner Verschachtelung zu komplex - siehe [Ticket 4266](https://github.com/kitodo/kitodo-production/issues/4266)
- UI: Verankerung einer übergreifenden Filtersuche auch jenseits der Aufgaben- und Vorgangsliste!? Wird die Filtersuche ggf. zu komplex und eine nachträgliche Facettierungsmöglichkeit wie bei der Trefferliste nach der Headersuche macht doch Sinn?
